### ISSUE TITLE
[ms-python.python] Missing "pythonFiles/lib/" directory in packaged extension

### ISSUE THREAD

**Issue by jankeromnes** [open-vsx/publish-extensions#89]
We currently auto-publish https://open-vsx.org/extension/ms-python/python from this repository with the following configuration:

https://github.com/open-vsx/publish-extensions/blob/31c5d426bf6370178cdea7ab65fd560ae9760903/extensions.json#L541-L547

Unfortunately, it seems that the uploaded extension is completely missing the `extension/pythonFiles/lib/` directory.

It would be nice to include this directory when publishing to OpenVSX, and to re-publish all currently broken versions of this extension.

**Comment by jankeromnes** [comment]
It seems that we need to install at least "debugpy" manually.

Maybe this "Local Build" section contains the right bootstrapping step(s) we're currently missing: https://github.com/microsoft/vscode-python/blob/master/CONTRIBUTING.md#local-build

**Comment by jankeromnes** [comment]
Should be fixed in https://github.com/open-vsx/publish-extensions/commit/7147dbe5917b9f28fc0a90aa3e2947e2039f4602

**Comment by jankeromnes** [comment]
Failed with:

```
Error: Command failed: python3 -m pip --disable-pip-version-check install -t ./pythonFiles/lib/python --no-cache-dir --implementation py --no-deps --upgrade -r requirements.txt && python3 -m pip --disable-pip-version-check install -r build/debugger-install-requirements.txt && python3 ./pythonFiles/install_debugpy.py && npm run clean && npm run package

Usage:   
  /usr/bin/python3 -m pip install [options] <requirement specifier> [package-index-options] ...
  /usr/bin/python3 -m pip install [options] -r <requirements file> [package-index-options] ...
  /usr/bin/python3 -m pip install [options] [-e] <vcs project url> ...
  /usr/bin/python3 -m pip install [options] [-e] <local project path> ...
  /usr/bin/python3 -m pip install [options] <archive url/path> ...

no such option: --implementation
```

Maybe let's just try without the `--implementation py` option (whatever that is).

**Comment by int19h** [comment]
The point of `--implementation py` is to ensure that the extension uses package wheels that are Python-implementation-agnostic. If not specified, it will use the wheels that match the Python implementation that you are using to run `pip`. 

And you need an up-to-date version of `pip` to support `--implementation`. If the package build is using system Python, it's likely that you're using `pip` packaged by your distro (e.g. `python3-pip` in Debian and Ubuntu), which can be extremely out of date. It's better to use a virtual environment for build tools, as described in the Setup section of the build guide.

**Comment by jankeromnes** [comment]
@int19h Many thanks for providing this explanation!

I'm not very familiar with Python or wheels -- would you happen to know a good sequence of commands that can be used, in our GitHub Actions workflow, just prior to publishing `ms-python.python` to OpenVSX, that would result in the best / cleanest / most recent extension package to be uploaded?

Currently we have this:

https://github.com/open-vsx/publish-extensions/blob/eaa3737b1225cf08771d4cf94c51d54e4c2a0013/extensions.json#L545

**Comment by int19h** [comment]
I'm not too familiar with the packaging myself, but this is how the VSIX gets built for CI:

https://github.com/microsoft/vscode-python/blob/master/.github/workflows/ci.yml#L146-L188

(You can probably simplify it a fair bit by inlining the `...-deps` jobs, and removing caching.)

**Comment by jankeromnes** [comment]
> I'm not too familiar with the packaging myself, but [this](https://github.com/microsoft/vscode-python/blob/2f8ff9abb88770e83741012edfc2ca16fe05b838/.github/workflows/ci.yml#L146-L188) is how the VSIX gets built for CI.

Many thanks! I guess the relevant part for us is [how all the various Python dependencies are installed](https://github.com/microsoft/vscode-python/blob/2f8ff9abb88770e83741012edfc2ca16fe05b838/.github/workflows/ci.yml#L37-L53):

```
# Set up some Python version (`env.PYTHON_VERSION`) with the GitHub Action `actions/setup-python@v2`, and then:
python -m pip install -U pip
python -m pip --disable-pip-version-check install -t ./pythonFiles/lib/python --no-cache-dir --implementation py --no-deps --upgrade -r requirements.txt
python -m pip install wheel
python -m pip --disable-pip-version-check install -r build/debugger-install-requirements.txt
python ./pythonFiles/install_debugpy.py
```

Also, for what it's worth, our [current command-line](https://github.com/open-vsx/publish-extensions/blob/eaa3737b1225cf08771d4cf94c51d54e4c2a0013/extensions.json#L545):

```
python3 -m pip --disable-pip-version-check install -t ./pythonFiles/lib/python --no-cache-dir --no-deps --upgrade -r requirements.txt &&
python3 -m pip --disable-pip-version-check install -r build/debugger-install-requirements.txt &&
python3 ./pythonFiles/install_debugpy.py
```

... failed again in tonight's publish run, with the output:

```
Error: Command failed: python3 -m pip --disable-pip-version-check install -t ./pythonFiles/lib/python --no-cache-dir --no-deps --upgrade -r requirements.txt && python3 -m pip --disable-pip-version-check install -r build/debugger-install-requirements.txt && python3 ./pythonFiles/install_debugpy.py && npm run clean && npm run package

Collecting future==0.18.2 (from -r requirements.txt (line 7))
  Downloading https://files.pythonhosted.org/packages/45/0b/38b06fd9b92dc2b68d58b75f900e97884c45bedd2ff83203d933cf5851c9/future-0.18.2.tar.gz (829kB)
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ModuleNotFoundError: No module named \'setuptools\'
    
    ----------------------------------------

Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-d419m6_1/future/
```

I guess we would also need a `pip install setuptools` somewhere, unless switching to the more official setup instructions already takes care of that.

**Comment by jankeromnes** [comment]
"Hopefully" fixes this issue, but let's see how our CI likes it (and whether the uploaded extension actually works).

**Comment by jankeromnes** [comment]
Seems to have worked! 🎉 ([source](https://github.com/open-vsx/publish-extensions/runs/893116704?check_suite_focus=true))

> 🚀  Published ms-python.python v2020.6.1

Also, the newly published `ms-python.python@2020.6.1` [on OpenVSX](https://open-vsx.org/extension/ms-python/python/2020.6.1) now seems to have the previously missing `extension/pythonFiles/lib/` directory, which is what we wanted to resolve this issue.

However, when comparing `ls -R` with what's published on Microsoft's Marketplace, we still see a few minor differences (notably a missing `extension/pythonFiles/lib/python/old_ptvsd`), but hopefully that's not too big of a problem:

```diff
@@ -283,18 +283,18 @@
 manager.js
 signal.d.ts
 signal.js
 widgetLoader.d.ts
 widgetLoader.js
 widgets.css

 ./out/ipywidgets/dist:
-0.ipywidgets.js
 1.ipywidgets.js
+2.ipywidgets.js
 ipywidgets.js

 ./pythonFiles:
 Notebooks intro.ipynb
 completion.py
 install_debugpy.py
 interpreterInfo.py
 lib
@@ -318,16 +318,17 @@

 ./pythonFiles/lib/python:
 debugpy
 future
 isort
 jedi
 libfuturize
 libpasteurize
+old_ptvsd
 parso
 past
 pyls_jsonrpc

 ./pythonFiles/lib/python/debugpy:
 ThirdPartyNotices.txt
 __init__.py
 __main__.py
```

**Comment by jankeromnes** [comment]
Meanwhile, @spoenemann or @jbicker, could you please delete every OpenVSX version of https://open-vsx.org/extension/ms-python/python except `2020.6.1`? This will allow me to re-publish fixed versions.

**Comment by spoenemann** [comment]
Done.

**Comment by int19h** [comment]
`old_ptvsd` missing will break 2020.6 if somebody tries to opt out of the new debugger (it was the old one, provided as a fallback). It's gone entirely in 2020.7, which is why the build script I linked to doesn't try to do anything about it.

But, given that 2020.7 is the most recent stable version of the extension, it's probably not worth spending cycles trying to fully fix your build of 2020.6 at this point.

**Comment by jankeromnes** [comment]
All published Python extensions should now work properly, and the publish config here is once again pointing to the latest available release:

https://github.com/open-vsx/publish-extensions/blob/1542b6ec2a98eb5fc153995866888744b5557b12/extensions.json#L540-L546

It should get published automatically tonight at 03:03 UTC.
